### PR TITLE
Feature/1750 std atomic

### PR DESCRIPTION
**Pull request by davidnich**
this pull request allows Qore to support atomic operations on all CPUs by using C++11 atomic support.

For example, we are building Qore for ARM* CPUs, and it's using Mutexes on those platforms now for atomic operations (i.e. it's very slow) - same with other CPU architectures not explicitly supported with assembly atomic operations.

This pull request does not change the ABI actually - the affected class is `QoreReferenceCounter`, which remains 4 bytes on all platforms with supported atomic operations - before: `sizeof(QoreReferenceCounter): 4`, after: `sizeof(QoreReferenceCounter): 4`.

This is because the actual reference count wrapped by `QoreReferenceCounter` remains 4 bytes.

API compatibility is slightly changed in that any code that directly accesses the `references` member must cast it to an `int` to compile (ex: `(int)ref->references`).

### CODE REVIEW

**Review by pvanek — APPROVED** [review]
the less custom code we are using, the better.

I hope it won't break anything.

**Comment by davidnich** [comment]
@pvanek I have been testing it, and everything is running without recompiling - so far it looks good - the majority of the changes in the pull request are deleting unnecessary custom code

File: lib/RSet.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ void RObject::setRSet(RSet* rs, int rcnt) {
    rcount = rcnt;
 #ifdef DEBUG
    if (rcount > references)
-      printd(0, "RObject::setRSet() this: %p '%s' cannot set rcount %d > references %d\n", this, getName(), rcount, references);
+      printd(0, "RObject::setRSet() this: %p '%s' cannot set rcount %d > references %d\n", this, getName(), rcount, (int)references);
```
Comment (omusil24) [inline lib/RSet.cpp]: I would use `references.load()` here to be consistent with the rest of the code. Same with the comparisons above and under this line.

Comment (davidnich) [reply lib/RSet.cpp]: ok done

**Pull request merged**